src/algebra/cubic_discriminant.lean

```diff
@@ -0,0 +1,277 @@
+
+variables {R : Type*} [semiring R] (P Q : cubic R)
+
+/-- Convert a cubic polynomial to a polynomial. -/
+def to_poly : polynomial R := C P.d + C P.c * X + C P.b * X ^ 2 + C P.a * X ^ 3
+
+lemma coeff_eq :
+  P.to_poly.coeff 3 = P.a ∧ P.to_poly.coeff 2 = P.b ∧ P.to_poly.coeff 1 = P.c
+    ∧ P.to_poly.coeff 0 = P.d :=
+by { rw [to_poly], simp only [coeff_add, coeff_C, coeff_C_mul, coeff_X, coeff_C_mul_X], norm_num }
+
+@[simp]
+theorem coeff_eq_a : P.to_poly.coeff 3 = P.a := (coeff_eq P).1
+
+@[simp]
+theorem coeff_eq_b : P.to_poly.coeff 2 = P.b := (coeff_eq P).2.1
+
+@[simp]
+theorem coeff_eq_c : P.to_poly.coeff 1 = P.c := (coeff_eq P).2.2.1
+
+@[simp]
+theorem coeff_eq_d : P.to_poly.coeff 0 = P.d := (coeff_eq P).2.2.2
+
+lemma eq_iff : P.to_poly = Q.to_poly ↔ P = Q :=
+begin
+  rcases ⟨P, Q⟩ with ⟨⟨_, _, _, _⟩, ⟨_, _, _, _⟩⟩,
+  split,
+  { rintro h,
+    simp only,
+    split,
+    { apply_fun (λ p, coeff p 3) at h,
+      rw [coeff_eq_a, coeff_eq_a] at h,
+      exact h },
+    { split,
+      { apply_fun (λ p, coeff p 2) at h,
+        rw [coeff_eq_b, coeff_eq_b] at h,
+        exact h },
+      { split,
+        { apply_fun (λ p, coeff p 1) at h,
+          rw [coeff_eq_c, coeff_eq_c] at h,
+          exact h },
+        { apply_fun (λ p, coeff p 0) at h,
+          rw [coeff_eq_d, coeff_eq_d] at h,
+          exact h } } } },
+  { intro h,
+    injection h with ha hb hc hd,
+    rw [ha, hb, hc, hd] }
+end
+
+lemma to_poly.injective : function.injective (to_poly : cubic R → polynomial R) :=
+λ P Q, (eq_iff P Q).mp
+
+theorem a_of_eq (h : P.to_poly = Q.to_poly) : P.a = Q.a :=
+by { rw [eq_iff] at h, cases P, cases Q, injection h }
+
+theorem b_of_eq (h : P.to_poly = Q.to_poly) : P.b = Q.b :=
+by { rw [eq_iff] at h, cases P, cases Q, injection h }
+
+theorem c_of_eq (h : P.to_poly = Q.to_poly) : P.c = Q.c :=
+by { rw [eq_iff] at h, cases P, cases Q, injection h }
+
+theorem d_of_eq (h : P.to_poly = Q.to_poly) : P.d = Q.d :=
+by { rw [eq_iff] at h, cases P, cases Q, injection h }
+
+end coeff
+
+----------------------------------------------------------------------------------------------------
+/-! ## Degree -/
+
+section degree
+
+variables {R : Type*} [semiring R] {P : cubic R}
+
+private lemma degree_lt (ha : P.a ≠ 0) :
+  (C P.d + C P.c * X + C P.b * X ^ 2).degree < (C P.a * X ^ 3).degree :=
+begin
+  rw [degree_C_mul_X_pow _ ha],
+  apply lt_of_le_of_lt (degree_add_le _ _),
+  rw [max_lt_iff],
+  split,
+  { apply lt_of_le_of_lt (degree_add_le _ _),
+    rw [max_lt_iff],
+    split,
+    { apply lt_of_le_of_lt degree_C_le,
+      exact dec_trivial },
+    { apply lt_of_le_of_lt (degree_C_mul_X_le _),
+      exact dec_trivial } },
+  { apply lt_of_le_of_lt (degree_C_mul_X_pow_le _ _),
+    exact dec_trivial }
+end
+
+theorem degree_eq_three (ha : P.a ≠ 0) : P.to_poly.degree = ↑3 :=
+by { rw [to_poly, degree_add_eq_right_of_degree_lt, degree_C_mul_X_pow _ ha], exact degree_lt ha }
+
+theorem ne_zero (ha : P.a ≠ 0) : P.to_poly ≠ 0 :=
```

Comment:
Again, this is quite a generic name. For the conclusion, it suffices if only 1 of the coefficients is nonzero. It doesn't have to be `a`. Maybe it makes more sense to define a predicate `ne_zero := P.a ≠ 0 ∨ P.b ≠ 0 ∨ P.c ≠ 0 ∨ P.d ≠ 0`. Then you can assume `P.ne_zero` in lemmas below.
On the other hand, I understand that `a ≠ 0` is the nondegenerate case. Still, I think someone in the future might thank you for supporting the degenerate cases as long as possible.